src/library.tests/WorkloadInformationProviderTests.cs

```diff
@@ -1,109 +1,174 @@
+            _workloadInformationProvider = _autoFake.Resolve<WorkloadInformationProvider>(TypedParameter.From(remoteContainerConnectionDetails));
+        }
+
+        [Theory]
+        [InlineData(1, 100)]
+        [InlineData(5, 20)]
+        [InlineData(10, 3)]
+        [InlineData(1, 3)]
+        public async void GetReachableServicesAsync_HeadlessService(int numServices, int numAddresses)
+        {
+            ConfigureHeadlessService(numServices: numServices, namingFunction: (i) => $"myapp-{i}", numAddresses: numAddresses, addressHostNamingFunction: (i) => $"Host-{i}");
+            var result = await _workloadInformationProvider.GetReachableEndpointsAsync(namespaceName: "", localProcessConfig: null, includeSameNamespaceServices: true, cancellationToken: default(CancellationToken));
+            // Doing numServices-1 when calculating because we are adding empty subset for one service and that will be skipped
+            Assert.Equal((numServices-1) * (numAddresses), result.Count());
+            foreach (var endpoint in result) {
+                if (endpoint.Ports.Any()) {
+                    Assert.Equal(endpoint.Ports.ElementAt(0).LocalPort, -1);
+                }
+            }
+            
+        }
+
+        [Theory]
+        [InlineData(5, 1)]
+        public async void GetReachableServicesAsync_PortsToIgnore_HeadlessService(int numServices, int numAddresses)
+        { 
+            // The ports provided to the service
+            List<int> portsOffered = new List<int> {300, 450, 180, 190, 312};
+            // The ports which should be ignored by the service
+            List<int> portsToIgnore = new List<int> {190, 300};
+            // The ports which were allocated by the service, these must not include any of the ignore ports
+            List<int> portsAllocated = new List<int> {450, 180, 312};
+            
+            ConfigureHeadlessService(numServices: numServices, namingFunction: (i) => $"myapp-{i}", numAddresses: numAddresses, addressHostNamingFunction: (i) => $"Host-{i}", "190, 300", portsOffered);           
+            var result = await _workloadInformationProvider.GetReachableEndpointsAsync(namespaceName: "", localProcessConfig: null, includeSameNamespaceServices: true, cancellationToken: default(CancellationToken));                       
+            List<int> portsInResult = new List<int>();
+            foreach (var endpoint in result)
+            {
+                if (endpoint.Ports.Any())
+                {
+                    // None of the ports to ignore must be used 
+                    Assert.Empty(endpoint.Ports.Where(p => portsToIgnore.Contains(p.RemotePort)));                    
+                    portsInResult.AddRange(endpoint.Ports.Select( p => p.RemotePort));
+                }
+            }
+            Assert.True(portsAllocated.All(p => portsInResult.Contains(p)));
```

Comment:
how does this check that portsInResult does not contain ignore ports? I think this line is making sure that all ports inside PortsAllocated are inside portsInResult, but if portsInresult has extra entries we wont catch it, will we?

Reply:
I'm checking this in line 70, which is making sure that non of the portsToIgnore appear in the results. 
I will be re-writing this to test the updated logic.